File: csharp/1544.make-the-string-great.cs
/*
 * @lc app=leetcode id=1544 lang=csharp
 *
 * [1544] Make The String Great
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {

    private const int Diff = (int)'a' - (int)'A';

    public string MakeGood(string s) {
        var list = new List<char>();
        foreach (char c in s) {
            if (list.Count == 0) {
                list.Add(c);
                continue;
            }
            char p = list[list.Count - 1];
            if (Math.Abs(c - p) == Diff) {
                list.RemoveAt(list.Count - 1);
            } else {
                list.Add(c);
            }
        }
        return new string(list.ToArray());
    }
}
// @lc code=end

File: java/1544.make-the-string-great.java
/*
 * @lc app=leetcode id=1544 lang=java
 *
 * [1544] Make The String Great
 */

// @lc code=start
class Solution {
    public String makeGood(String s) {
        int n = s.length();
        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; ++i) {
            char c = s.charAt(i);
            if (Character.isUpperCase(c) && sb.length() > 0 && sb.charAt(sb.length() - 1) == Character.toLowerCase(c)) {
                sb.deleteCharAt(sb.length() - 1);
            } else if (Character.isLowerCase(c) && sb.length() > 0 && sb.charAt(sb.length() - 1) == Character.toUpperCase(c)) {
                sb.deleteCharAt(sb.length() - 1);
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }
}
// @lc code=end

